File: tyhp-lang/vscode/src/debug/proxyLifecycle.ts
/**
 * Pure start/stop state machine for the XDebug proxy child process.
 * Unexpected exit from `running` is `error`; a requested stop ends in `stopped`.
 */

export type ProxyRunState = "stopped" | "starting" | "running" | "stopping" | "error";

export type ProxyLifecycleEvent =
    | { type: "startRequested" }
    | { type: "listening" }
    | { type: "startFailed" }
    | { type: "stopRequested" }
    | { type: "exited"; expected: boolean };

export function nextProxyState(state: ProxyRunState, event: ProxyLifecycleEvent): ProxyRunState {
    switch (event.type) {
        case "startRequested":
            if (state === "running" || state === "starting") {
                return state;
            }
            return "starting";
        case "listening":
            return state === "starting" || state === "running" ? "running" : state;
        case "startFailed":
            return state === "starting" ? "error" : state;
        case "stopRequested":
            if (state === "stopped") {
                return "stopped";
            }
            return "stopping";
        case "exited":
            if (state === "error") {
                return "error";
            }
            if (event.expected || state === "stopping" || state === "stopped") {
                return "stopped";
            }
            return "error";
        default:
            return state;
    }
}

export function proxyIsActive(state: ProxyRunState): boolean {
    return state === "starting" || state === "running" || state === "stopping";
}

export function proxyIsListening(state: ProxyRunState): boolean {
    return state === "running";
}

File: tyhp-lang/phpstorm/src/main/kotlin/com/tyhp/lang/debug/ProxyLifecycle.kt
package com.tyhp.lang.debug

/**
 * Pure start/stop state machine for the XDebug proxy child process.
 * Unexpected exit from `running` is `error`; a requested stop ends in `stopped`.
 */

enum class ProxyRunState {
    STOPPED,
    STARTING,
    RUNNING,
    STOPPING,
    ERROR,
}

sealed class ProxyLifecycleEvent {
    data object StartRequested : ProxyLifecycleEvent()
    data object Listening : ProxyLifecycleEvent()
    data object StartFailed : ProxyLifecycleEvent()
    data object StopRequested : ProxyLifecycleEvent()
    data class Exited(val expected: Boolean) : ProxyLifecycleEvent()
}

fun nextProxyState(state: ProxyRunState, event: ProxyLifecycleEvent): ProxyRunState {
    return when (event) {
        ProxyLifecycleEvent.StartRequested ->
            if (state == ProxyRunState.RUNNING || state == ProxyRunState.STARTING) {
                state
            } else {
                ProxyRunState.STARTING
            }
        ProxyLifecycleEvent.Listening ->
            if (state == ProxyRunState.STARTING || state == ProxyRunState.RUNNING) {
                ProxyRunState.RUNNING
            } else {
                state
            }
        ProxyLifecycleEvent.StartFailed ->
            if (state == ProxyRunState.STARTING) ProxyRunState.ERROR else state
        ProxyLifecycleEvent.StopRequested ->
            if (state == ProxyRunState.STOPPED) {
                ProxyRunState.STOPPED
            } else {
                ProxyRunState.STOPPING
            }
        is ProxyLifecycleEvent.Exited -> {
            if (state == ProxyRunState.ERROR) {
                ProxyRunState.ERROR
            } else if (event.expected || state == ProxyRunState.STOPPING || state == ProxyRunState.STOPPED) {
                ProxyRunState.STOPPED
            } else {
                ProxyRunState.ERROR
            }
        }
    }
}

fun proxyIsActive(state: ProxyRunState): Boolean =
    state == ProxyRunState.STARTING ||
        state == ProxyRunState.RUNNING ||
        state == ProxyRunState.STOPPING

fun proxyIsListening(state: ProxyRunState): Boolean = state == ProxyRunState.RUNNING
